rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,323 @@
+
+#### Pros:
+
+*   Simplified graph processing since the operation is just a single Node in the
+    graph.
+*   Clean graph def.
+*   Tooling (once supports new ops) stays forward compatible. The underlying
+    implementation is invisible.
+
+#### Cons:
+
+Introduces backward incompatible changes that could be avoided. Every time an
+operation is added, the versions of Tensorflow that would have been capable of
+processing the original graph (with core ops), will no longer be able to read
+the graph def that contains the new to process the new version, even though the
+only thing that *has* changed is the tensorflow added a faster implementation of
+that Op.
+
+For example: consider matrix exponentiation. It is implemented as a fairly
+complex tensorflow graph that uses just regular matrix multiplication and other
+standard ops. However, one can easily imagine that this implementation could be
+highly optimized if done as a single expm node, however if we replace that, it
+will break old tools, which would either need to effectively re-implement
+tensorflow original implementation.
+
+Requires custom version of tensorflow to add new ops outside of tensorflow which
+makes it out of reach for most users and introduces incompatibilities within the
+ecosystem, effectively forcing large users to be stuck at old versions of tensorflow.
+
+#### Alternative 2: Use name scopes for delineation of the ops
+
+Pros:
+
+*   Simple and intuitive graph-def
+*   Backward compatible - no new operations are added.
+
+Cons:
+
+*   very brittle for the tools to optimize such ops. If they depend on scope
+    names it can easily cause conflicts with models that are doing something
+    else or accidental renames can cause scopes to become invisible for the
+    graph processing.
+*   If tools depend on graph pattern matching, this makes it hard to change
+    implementations later on.
+
+*   Tooling is not forward compatible.
+
+## Questions and Discussion Topics
+
+1.  We should provide namespacing for the operation names early on to avoid
+    future collisions with function names. One option is to adopt java-style
+    "org.opname", basically using the developing organization as disambiguating
+    factor. Another alternative is to use semantic namespaces e.g. `linalg`.  Yet
+    third, is to use org.semantic.opname.
+    
+    Note: since many of these functions will live in user codebase we obviously
+    can't and shouldn't police them, however having a recommended style guide
+    will simplify things later on.
+    
+    Should there be a `graduation` process? Where ops eventually move to a higher tier.
+    E.g. dedicated top-tier op like tensorflow.opname?
+    If so, we might also consider having another attribute 'aliases' to allow
+    history of names for backward compatibility.
+
+2.  If we go org/name, do we need centralized registry of accepted "org" names?
+
+3.  Do we need `reference` field? The isdea is that points to the source of
```

Comment:
right, the idea was that automatic tools to rely on it as human-readable hash (or use non-tensorflow implementation s s a source of groundtruth), but perhaps this would be too confusing.

Reply:
I do like the idea, but I'm also in favor of just keeping references in doc string. If a downstream developer is re-implementing a composite op I think they'll be thoroughly reading the doc.